django/chat/views.py

```diff
@@ -706,20 +707,79 @@ def set_security_label(request, chat_id, security_label_id):
     )
 
 
+def highlight_claims(claims_list, text, threshold=80):
+    # match if the claims_list exist is text; if it does, then highlight it with  <mark> tag
+    from langchain_text_splitters import RecursiveCharacterTextSplitter
+    from rapidfuzz import fuzz
+
+    text_splitter = RecursiveCharacterTextSplitter(
+        chunk_size=100,
+        chunk_overlap=20,
+        length_function=len,
+        is_separator_regex=False,
+    )
+
+    good_matches = []
+    # Split the text into chunks
+    chunks = text_splitter.create_documents([text])
+
+    for claim in claims_list:
+        for chunk in chunks:
+            chunk_text = chunk.page_content
+            # Find fuzzy matches
+            score = fuzz.partial_ratio(chunk_text.lower(), claim.lower())
+            if score >= threshold:
+                good_matches.append(chunk_text)
+
+    for match in good_matches:
+        if len(match) > 3:
+            text = text.replace(match, f"<mark>{match}</mark>")
+
+    return text
+
+
+from chat.llm import OttoLLM
+
+
+def extract_claims_from_llm(llm_response_text):
+    llm = OttoLLM()
+    prompt = f"""
+    Based on the following LLM response, extract key factual claims, including direct quotes.
+
+    Respond in the format:
+    <claim>whatever the claim is...</claim>
+    <claim>another claim...</claim>
+
+    etc.
+
+    ---
+    <llm_response>
+    {llm_response_text}
+    </llm_response>
+    """
+    claims_response = llm.complete(prompt)
+    # find the claim tags and add whats wrapped in the claim tags to a list
+    claims_list = re.findall(r"<claim>(.*?)</claim>", claims_response)
+    return claims_list
+
+
 @permission_required("chat.access_message", objectgetter(Message, "message_id"))
 def message_sources(request, message_id):
     sources = []
 
     for source in AnswerSource.objects.prefetch_related(
         "document", "document__data_source", "document__data_source__library"
     ).filter(message_id=message_id):
+
         source_text = str(source.node_text)
```

Comment:
Because `source.node_text` is actually a function, you can update that function to do the claim highlighting (if no highlighted_text property) or just return the pre-highlighted text.